File: plugins/robots/robotsGenerator/nxtOSEK/generators/simpleElements/balanceGenerator.cpp
#include "balanceGenerator.h"
#include "../../nxtOSEKRobotGenerator.h"

using namespace robots::generator;

BalanceGenerator::BalanceGenerator()
{
}

QList<SmartLine_old> BalanceGenerator::convertElementIntoDirectCommand(NxtOSEKRobotGenerator *nxtGen
		, qReal::Id const &elementId, qReal::Id const &logicElementId)
{
	QList<SmartLine_old> result;

	QString const forward = nxtGen->api()->stringProperty(logicElementId, "forward");
	QString const turn = nxtGen->api()->stringProperty(logicElementId, "turn");
	QString const gyroPort = nxtGen->api()->stringProperty(logicElementId, "port");
	QString const gyroValue = nxtGen->api()->stringProperty(logicElementId, "gyroOffset");
	QString const firstMotorValue = nxtGen->api()->stringProperty(logicElementId, "port1");
	QString const secondMotorValue = nxtGen->api()->stringProperty(logicElementId, "port2");
	QString const outPowerValue1 = nxtGen->api()->stringProperty(logicElementId, "pwm1");
	QString const outPowerValue2 = nxtGen->api()->stringProperty(logicElementId, "pwm2");
	result << (SmartLine_old(QString("balance_control("), elementId)) << (SmartLine_old(QString("(F32)") + forward + ", "
			, elementId)) << (SmartLine_old(QString("(F32)") + turn + ", ", elementId))
			<< (SmartLine_old(QString("(F32)ecrobot_get_gyro_sensor(NXT_PORT_S") + gyroPort + "), ", elementId))
			<< (SmartLine_old(QString("(F32)") + gyroValue + ", ", elementId))
			<< (SmartLine_old(QString("(F32)nxt_motor_get_count(NXT_PORT_") + firstMotorValue + "), ", elementId))
			<< (SmartLine_old(QString("(F32)nxt_motor_get_count(NXT_PORT_") + secondMotorValue + "), ", elementId))
			<< (SmartLine_old(QString("(F32)ecrobot_get_battery_voltage(), "), elementId))
			<< (SmartLine_old(QString("&") + outPowerValue1 + ", ", elementId))
			<< (SmartLine_old(QString("&") + outPowerValue2 + ");", elementId));

	nxtGen->activateBalancer();

	return result;
}

File: plugins/visualInterpreter/examples/robotsCodeGeneration/reactionsStorage/BalanceGenerator.py
# Application condition
balance.id == max_used_id and not cur_node_is_processed

# Reaction
balance_code = "balance_control((F32)" + balancer.forward + ", (F32)" + balancer.turn + ", (F32)ecrobot_get_gyro_sensor(NXT_PORT_S" + balancer.port + "), (F32)" + balancer.gyroOffset + ", (F32)nxt_motor_get_count(NXT_PORT_" + balancer.port1 + "), (F32)nxt_motor_get_count(NXT_PORT_" + balancer.port2 + "), (F32)ecrobot_get_battery_voltage(), &" + balancer.pwm1 + ", &" + balancer.pwm2 + ");\n"

code.append([balance_code])
id_to_pos_in_code[balancer.id] = len(code) - 1

cur_node_is_processed = True
